src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProducerExtension.cs

```diff
@@ -184,7 +191,7 @@ internal Task DeliverItem(StreamId streamId, object item, bool fireAndForgetDeli
             {
                 var tasks = fireAndForgetDelivery ? null : new List<Task>();
                 var immutableItem = new Immutable<object>(item);
-                foreach (KeyValuePair<GuidId, Tuple<IStreamConsumerExtension, IStreamFilterPredicateWrapper>> subscriptionKvp in consumers)
+                foreach (KeyValuePair<GuidId, Tuple<IStreamConsumerExtension, IStreamFilterPredicateWrapper>> subscriptionKvp in consumers.ToList())
```

Comment:
I think you can not do it this way, even in pub sub grain, since await can be inlined and execute in the same turn, in theory.
I think you need to really Chaney the code flow here and in the pub sub grain, to be for sure safe. 
Let me know if you agree (that we need tom change that) and if you are not sure how, we can think together. I don't immediately now know how, but it can be some kind of variation of return value, that is awaited and guarantees all removals happen only after all calls were made, but also the overall loop is not globally awaited.